File: solutions/fortran/protein-translation/1/protein_translation.f90
module protein_translation
  implicit none

contains

  function proteins(rna) result(names)
    character(len=*), intent(in) :: rna
    character(len=13), allocatable :: names(:)
    character(len=13) :: cod
    integer :: i, first_stop
    
    first_stop=0
    allocate(names(len(rna)/3))
      do i=1,len(rna),3
        cod=translation(rna(i:i+2))
        if (cod=='STOP') then
          first_stop=((i-1)+3)/3
        end if
        names(((i-1)+3)/3)=cod
      end do
      if (first_stop/=0) then
        names=names(1:first_stop-1)
      end if
  end function proteins

  function translation(codon) result (match)
    character(len=13) :: match
    character(len=3) :: codon
   
    select case (codon)
        case ("AUG")
            match="Methionine"
        case ("UGG")
            match="Tryptophan"
        case ("UUU", "UUC")
            match="Phenylalanine"
        case ("UUA", "UUG")
            match="Leucine"
        case ("UAU", "UAC")
            match="Tyrosine"
        case ("UGU", "UGC")
            match="Cysteine"
        case ("UCU", "UCC","UCA", "UCG")
            match="Serine"
        case ("UAA", "UAG", "UGA")
            match="STOP"
        case default
            match=""
    end select
  end function translation

end module protein_translation

File: solutions/go/protein-translation/1/protein_translation.go
package proteintranslation

import (
	"errors"
    "fmt"
)

var ErrStop = errors.New("stop codon encountered")
var ErrInvalidBase = errors.New("invalid base")

var m map[string]string

func FromRNA(rna string) ([]string, error) {
	var error_msg error
	codon := ""
	var codons []string
	for i := 0; i < len(rna); i = i + 3 {
		if i+3 > len(rna) {
			codon = rna[i : len(rna)-1]
		} else {
			codon = rna[i : i+3]
		}
		value, err := FromCodon(codon)
        fmt.Println(codon, value, err)
        if err == ErrStop {
            return codons, nil
        } else if err == nil {
			codons = append(codons, value)
		} else {
			error_msg = err
		}
	}
	if error_msg==ErrStop || error_msg == nil {
    	return codons, nil    
    } else {
        return codons, error_msg
    }
}

func FromCodon(codon string) (string, error) {
	if codon == "UAA" || codon == "UAG" || codon == "UGA" {
		return "", ErrStop
	} else {
		m = make(map[string]string)
		m["AUG"] = "Methionine"
		m["UUU"] = "Phenylalanine"
		m["UUC"] = "Phenylalanine"
		m["UUA"] = "Leucine"
		m["UUG"] = "Leucine"
		m["UAU"] = "Tyrosine"
		m["UAC"] = "Tyrosine"
		m["UGU"] = "Cysteine"
		m["UGC"] = "Cysteine"
		m["UGG"] = "Tryptophan"
		m["UCU"] = "Serine"
		m["UCC"] = "Serine"
		m["UCA"] = "Serine"
		m["UCG"] = "Serine"
		elem, ok := m[codon]
		if ok == false {
			return "", ErrInvalidBase
		} else {
			return elem, nil
		}
	}
}
